File: api/routes/chats.js
const express = require('express');
const router = express.Router();
const mongoose = require('mongoose');
const { ClerkExpressRequireAuth } = require("@clerk/clerk-sdk-node");
require('dotenv').config();
const Message = require('../models/Message');
const Room = require('../models/Room');
const User = require('../models/User');
// ✅ Call the Clerk middleware to get a handler function
const clerkAuth = ClerkExpressRequireAuth();

// GET /api/chats
router.get('/list',clerkAuth, async (req, res) => {

  console.log('$$$$..process.env',process.env.CLERK_SECRET_KEY);
  

   const myId = req.auth.userId;

   console.log('called profile..myID',myId);
    try {
    // ✅ 1️⃣ 1-to-1 chats
    const oneToOne = await Message.aggregate([
      {
        $match: {
          $or: [
            { senderId: myId },
            { receiverId: myId }
          ],
          roomId: { $exists: false }
        }
      },
      { $sort: { createdAt: -1 } },
      {
        $group: {
          _id: {
            user: {
              $cond: [
                { $eq: ["$senderId", myId] },
                "$receiverId",
                "$senderId"
              ]
            }
          },
          lastMessage: { $first: "$$ROOT" }
        }
      },
      {
        $lookup: {
          from: "users",
          localField: "_id.user",
          foreignField: "clerkId",
          as: "contact"
        }
      },
      { $unwind: "$contact" },
      {
        $project: {
          _id: 0,
          userId: "$_id.user",          
          lastMessageText: "$lastMessage.text",
          lastMessageTime: "$lastMessage.createdAt",
          receiverId:"$lastMessage.receiverId",
          contactName: "$contact.name",
          phoneNumber:"$contact.phoneNumber",
          contactPhoto: "$contact.photoUrl",
          contactStatus: "$contact.status"
        }
      }
    ]);

    // ✅ 2️⃣ Rooms
    const rooms = await Message.aggregate([
      {
        $match: {
          roomId: { $exists: true }
        }
      },
      { $sort: { createdAt: -1 } },
      {
        $group: {
          _id: "$roomId",
          lastMessage: { $first: "$$ROOT" }
        }
      },
      {
        $lookup: {
          from: "rooms",
          localField: "_id",
          foreignField: "_id",
          as: "room"
        }
      },
      { $unwind: "$room" },
      {
        $project: {
          _id: 0,
          roomId: "$_id",
          lastMessageText: "$lastMessage.text",
          lastMessageTime: "$lastMessage.createdAt",
          photoUrl:"$room.roomPic",
          roomName: "$room.name",
          roomMembers: "$room.members"
        }
      }
    ]);
   console.error('oneToOne and rooms',oneToOne);
    res.json({ oneToOne, rooms });
  } catch (err) {
    console.error(err);
    res.status(500).json({ message: "Server error" });
  }
});

module.exports = router;

File: api/routes/chats.py
from fastapi import APIRouter, HTTPException, Depends
from fastapi_clerk_auth import ClerkConfig, ClerkHTTPBearer, HTTPAuthorizationCredentials
from motor.motor_asyncio import AsyncIOMotorClient
import os

router = APIRouter()

# ✅ MongoDB connection
client = AsyncIOMotorClient(os.getenv("MONGO_URI", "mongodb+srv://skyees:raj1234%23A@skyees.oabwff3.mongodb.net/skyees?retryWrites=true&w=majority"))
db = client["skyees"]
messages_collection = db["messages"]

# ✅ Clerk config (validates JWTs from Authorization header)
clerk_config = ClerkConfig(
    jwks_url="https://clerk.skyees.com/.well-known/jwks.json"
)
clerk_auth_guard = ClerkHTTPBearer(config=clerk_config)

@router.get("/list")
async def get_chats(credentials: HTTPAuthorizationCredentials = Depends(clerk_auth_guard)):
    print("📡 /api/chats/list called by:", credentials.decoded.get("sub")) # Added log
    try:
        myId = credentials.decoded.get("sub")

        oneToOne = await messages_collection.aggregate([
            {
                "$match": {
                    "$or": [{"senderId": myId}, {"receiverId": myId}],
                    "roomId": {"$exists": False}
                }
            },
            {"$sort": {"createdAt": -1}},
            {
                "$group": {
                    "_id": {
                        "$cond": [
                            {"$eq": ["$senderId", myId]},
                            "$receiverId",
                            "$senderId"
                        ]
                    },
                    "lastMessage": {"$first": "$$ROOT"}
                }
            },
            {
                "$lookup": {
                    "from": "users",
                    "localField": "_id", # This is the other user's ID
                    "foreignField": "clerkId",
                    "as": "contact"
                }
            },
            {"$unwind": {"path": "$contact", "preserveNullAndEmptyArrays": True}},
            {
                "$project": {
                    "_id": 0,
                    "userId": "$_id",
                    "lastMessageText": "$lastMessage.text",
                    "lastMessageTime": "$lastMessage.createdAt",
                    "contactName": {"$ifNull": ["$contact.username", "Unknown User"]},
                    "contactPhoto": "$contact.profilePic"
                }
            }
        ]).to_list(length=None)

        return {"oneToOne": oneToOne, "rooms": []} # Simplified for debugging
    except Exception as e:
        print("❌ Error in /api/chats/list:", str(e))
        raise HTTPException(status_code=500, detail="Server error fetching chats")